- #### Download data from here:<br>
https://github.com/cwida/duckdb-data/releases/download/v1.0/taxi_2019_04.parquet <br>
https://github.com/cwida/duckdb-data/releases/download/v1.0/taxi_2019_05.parquet <br>
https://github.com/cwida/duckdb-data/releases/download/v1.0/taxi_2019_06.parquet


### **Install Python Libraries**


#### Run this command in your python (virtual) environment


#### <i> `pip install duckdb magic_duckdb polars plotly_express nbformat --quiet --user` </i>


#### 1. duckdb:

DuckDB is an embedded analytical SQL database for Python.
It is designed for fast query execution and low memory usage.
Developers often use it for data analysis, data manipulation, and reporting.
You can find more information about DuckDB on their official website .

#### 2. magic_duckdb:

magic_duckdb is a Python package that provides Jupyter Notebook magic commands for interacting with DuckDB.
It allows you to run SQL queries against DuckDB directly within a Jupyter Notebook.
You can explore its usage and documentation in the GitHub repository: <br> https://github.com/iqmo-org/magic_duckdb, or here:<br>https://pypi.org/project/magic-duckdb/

#### 3. Polars:

Polars is a fast DataFrame library for Python and Rust.
It is designed for big data processing and provides a similar interface to Pandas.
Polars is particularly useful for working with large datasets efficiently.
To learn more about Polars visit the official website:<br> https://pola.rs/

#### 4. Plotly Express:

Plotly Express is a high-level Python visualization library built on top of Plotly.
It simplifies the creation of interactive plots, charts, and graphs.
With Plotly Express, you can quickly generate visualizations without writing extensive code.
Explore its capabilities in the official documentation: <br>https://plotly.com/python/plotly-express/

#### 5. nbformat:

nbformat is a Python library for working with Jupyter Notebook file formats.
It allows you to read, write, and manipulate Jupyter Notebook files programmatically.
Developers often use it for tasks like converting notebooks to different formats or extracting cell content.
You can find more details in the official documentation: <br>https://nbformat.readthedocs.io/en/latest/ and here: <br>https://pypi.org/project/nbformat/


#### Use the `--quiet` flag to suppress output during installation, and the `--user` flag to install the packages in your user-specific Python environment.


In [1]:
# .torenv\Scripts\Activate.ps1

In [2]:
import duckdb
import pandas as pd
import os
import sys
#import plotly_express as px

%load_ext magic_duckdb

#### We already installed and loaded the duckdb magic in our notebook. Let us take advantage of it <br>so that we don't repeate `duckdb.sql` ever time. 

<br>Instead we can use: <br>
- `%dql` for single line queries and:<br>
- `%%dql` for multi-line queries instead


##### Because we are using the magic_duckdb extension, our queries will return a Pandas DataFrame, <br> bringing the entire query result into memory.

* We can avoid this by setting the type of return by using `"-t"` followed by the type, choosing from "df", "arrow", "pl", "describe", "show" and "relation".


#### We can also query files over the internet using duckdb's `httpfs` extension

In [3]:
%%dql -t df
INSTALL httpfs;

Empty DataFrame
Columns: [Success]
Index: []

In [4]:
%%dql -t df
LOAD  httpfs;

Empty DataFrame
Columns: [Success]
Index: []

In [5]:
%%dql
SELECT format('{:,}', COUNT(*)) as count 
FROM 'https://github.com/cwida/duckdb-data/releases/download/v1.0/taxi_2019_04.parquet'

count
0  7,433,139

In [6]:
%%time
%%dql -t show
SELECT * FROM 'https://github.com/cwida/duckdb-data/releases/download/v1.0/taxi_2019_04.parquet' LIMIT 5;

┌───────────┬─────────────────────┬─────────────────────┬───┬──────────────┬──────────────────────┐
│ vendor_id │      pickup_at      │     dropoff_at      │ … │ total_amount │ congestion_surcharge │
│  varchar  │      timestamp      │      timestamp      │   │    float     │        float         │
├───────────┼─────────────────────┼─────────────────────┼───┼──────────────┼──────────────────────┤
│ 1         │ 2019-04-01 00:04:09 │ 2019-04-01 00:06:35 │ … │          8.8 │                  2.5 │
│ 1         │ 2019-04-01 00:22:45 │ 2019-04-01 00:25:43 │ … │          8.3 │                  2.5 │
│ 1         │ 2019-04-01 00:39:48 │ 2019-04-01 01:19:39 │ … │        47.75 │                  2.5 │
│ 1         │ 2019-04-01 00:35:32 │ 2019-04-01 00:37:11 │ … │          7.3 │                  2.5 │
│ 1         │ 2019-04-01 00:44:05 │ 2019-04-01 00:57:58 │ … │        23.15 │                  2.5 │
├───────────┴─────────────────────┴─────────────────────┴───┴──────────────┴──────────────────────┤


#### Let us inspect the internet parquet file from our local duckdb using the httpfs extension


In [7]:
%%time
%%dql
SELECT COUNT(*) AS RowCount, 
       AVG(passenger_count) AS avg_number_of_passengers,
       AVG(trip_distance) AS avg_trip_distance,
       MAX(trip_distance) AS max_trip_distance,
       AVG(fare_amount) AS avg_fare_amount,
       MAX(fare_amount) AS max_fare_amount,
       AVG(tip_amount) AS avg_tip_amount,
       MAX(tip_amount) AS max_tip_amount FROM 'https://github.com/cwida/duckdb-data/releases/download/v1.0/taxi_2019_04.parquet';

CPU times: total: 2.38 s
Wall time: 11.4 s


RowCount  avg_number_of_passengers  avg_trip_distance  max_trip_distance  \
0   7433139                  1.573301           2.998019         830.900024   

   avg_fare_amount  max_fare_amount  avg_tip_amount  max_tip_amount  
0         13.19433      395839.9375        2.212688      440.799988

#### Find current working directory path

In [8]:
%pwd 

'c:\\Users\\hamin\\Desktop\\Learn-DuckDB\\duck-db-notebooks'

In [9]:
# Get the current working directory
current_dir = os.getcwd()
# Move up one level from the current directory
parent_dir = os.path.dirname(current_dir)
# Change directory into db directory
data_dir = os.path.join(parent_dir, 'data')

In [10]:
file_path = os.path.join(data_dir, 'taxi_2019_04.parquet')

#### Let us create a local copy of the internet taxi_2019_04 parquet file

In [11]:
file_path

'c:\\Users\\hamin\\Desktop\\Learn-DuckDB\\data\\taxi_2019_04.parquet'

In [12]:
%%time
%%dql
COPY (SELECT * FROM READ_PARQUET('https://github.com/cwida/duckdb-data/releases/download/v1.0/taxi_2019_04.parquet'))
      TO 'data/taxi_2019_04.parquet' (FORMAT 'parquet');

CPU times: total: 7.78 s
Wall time: 17.5 s


Count
0  7433139

#### Let us verify rowcount in the newly created parquet file

In [13]:
%%time
%%dql
SELECT format('{:,}', COUNT(*)) as count FROM 'data/taxi_2019_04.parquet';

CPU times: total: 31.2 ms
Wall time: 46.7 ms


#### We can also exclude columns using an matching expression

In [14]:
%%dql -t df
SELECT COLUMNS(c -> c LIKE 'to%') FROM 'data/taxi_2019_04.parquet' limit 5;

tolls_amount  total_amount
0           0.0          8.80
1           0.0          8.30
2           0.0         47.75
3           0.0          7.30
4           0.0         23.15

#### Let us look at the datatypes in the parquet file - <i>Note the duckdb sql syntax </i>

In [15]:
%%dql
DESCRIBE FROM 'data/taxi_2019_04.parquet';

column_name column_type null   key default extra
0               vendor_id     VARCHAR  YES  None    None  None
1               pickup_at   TIMESTAMP  YES  None    None  None
2              dropoff_at   TIMESTAMP  YES  None    None  None
3         passenger_count     TINYINT  YES  None    None  None
4           trip_distance       FLOAT  YES  None    None  None
5            rate_code_id     VARCHAR  YES  None    None  None
6      store_and_fwd_flag     VARCHAR  YES  None    None  None
7      pickup_location_id     INTEGER  YES  None    None  None
8     dropoff_location_id     INTEGER  YES  None    None  None
9            payment_type     VARCHAR  YES  None    None  None
10            fare_amount       FLOAT  YES  None    None  None
11                  extra       FLOAT  YES  None    None  None
12                mta_tax       FLOAT  YES  None    None  None
13             tip_amount       FLOAT  YES  None    None  None
14           tolls_amount       FLOAT  YES  None    None  None
15  improvement_surcharge       FLOAT  YES  None    None  None
16           total_amount       FLOAT  YES  None    None  None
17   congestion_surcharge       FLOAT  YES  None    None  None

### Let us lok at the Parquet file


#### Let us take a look at the parquet file's metadata

In [16]:
%%dql 
SELECT *
FROM parquet_metadata('data/taxi_2019_04.parquet')

file_name  row_group_id  row_group_num_rows  \
0     data/taxi_2019_04.parquet             0              122880   
1     data/taxi_2019_04.parquet             0              122880   
2     data/taxi_2019_04.parquet             0              122880   
3     data/taxi_2019_04.parquet             0              122880   
4     data/taxi_2019_04.parquet             0              122880   
...                         ...           ...                 ...   
1093  data/taxi_2019_04.parquet            60               60339   
1094  data/taxi_2019_04.parquet            60               60339   
1095  data/taxi_2019_04.parquet            60               60339   
1096  data/taxi_2019_04.parquet            60               60339   
1097  data/taxi_2019_04.parquet            60               60339   

      row_group_num_columns  row_group_bytes  column_id  file_offset  \
0                        18         15636480          0            0   
1                        18         15636480          1            0   
2                        18         15636480          2            0   
3                        18         15636480          3            0   
4                        18         15636480          4            0   
...                     ...              ...        ...          ...   
1093                     18          7818240         13            0   
1094                     18          7818240         14            0   
1095                     18          7818240         15            0   
1096                     18          7818240         16            0   
1097                     18          7818240         17            0   

      num_values         path_in_schema        type  ...      stats_min_value  \
0         122880              vendor_id  BYTE_ARRAY  ...                    1   
1         122880              pickup_at       INT64  ...  2019-03-31 09:30:05   
2         122880             dropoff_at       INT64  ...  2019-03-31 14:02:34   
3         122880        passenger_count       INT32  ...                    0   
4         122880          trip_distance       FLOAT  ...                  0.0   
...          ...                    ...         ...  ...                  ...   
1093       60339             tip_amount       FLOAT  ...                -1.56   
1094       60339           tolls_amount       FLOAT  ...                -6.12   
1095       60339  improvement_surcharge       FLOAT  ...                 -0.3   
1096       60339           total_amount       FLOAT  ...                -59.8   
1097       60339   congestion_surcharge       FLOAT  ...                 -2.5   

          stats_max_value  compression              encodings  \
0                       4       SNAPPY  PLAIN, RLE_DICTIONARY   
1     2019-04-01 23:59:46       SNAPPY                  PLAIN   
2     2019-04-02 15:41:09       SNAPPY                  PLAIN   
3                       6       SNAPPY                  PLAIN   
4                   202.7       SNAPPY                  PLAIN   
...                   ...          ...                    ...   
1093                315.0       SNAPPY                  PLAIN   
1094                86.12       SNAPPY                  PLAIN   
1095                  0.3       SNAPPY                  PLAIN   
1096               330.96       SNAPPY                  PLAIN   
1097                  2.5       SNAPPY                  PLAIN   

     index_page_offset dictionary_page_offset data_page_offset  \
0                  NaN                    4.0               34   
1                  NaN                    NaN            30691   
2                  NaN                    NaN           627539   
3                  NaN                    NaN          1252963   
4                  NaN                    NaN          1359382   
...                ...                    ...              ...   
1093               NaN                    NaN        196428794   
1094               NaN                    NaN 

#### Let us limit the number of columns to what we are looking for

In [17]:
%%dql 
SELECT file_name, total_compressed_size, total_uncompressed_size 
FROM parquet_metadata('data/taxi_2019_04.parquet')

file_name  total_compressed_size  \
0     data/taxi_2019_04.parquet                  30687   
1     data/taxi_2019_04.parquet                 596848   
2     data/taxi_2019_04.parquet                 625424   
3     data/taxi_2019_04.parquet                 106419   
4     data/taxi_2019_04.parquet                 316613   
...                         ...                    ...   
1093  data/taxi_2019_04.parquet                 127030   
1094  data/taxi_2019_04.parquet                  18861   
1095  data/taxi_2019_04.parquet                  11703   
1096  data/taxi_2019_04.parquet                 149845   
1097  data/taxi_2019_04.parquet                  25435   

      total_uncompressed_size  
0                       59236  
1                      983071  
2                      983071  
3                      491551  
4                      491551  
...                       ...  
1093                   241387  
1094                   241387  
1095                   241387  
1096                   241387  
1097                   241387  

[1098 rows x 3 columns]

#### Let us see how much space the parquet file takes up on disk

In [18]:
%%dql -t df
SELECT 
       format('{:,}', CAST(ROUND((sum(total_compressed_size) / (1024))) AS INT)) as compressed_KB,
       format('{:,}', CAST(ROUND((sum(total_uncompressed_size) / (1024))) AS INT)) as uncompressed_KB,
       format('{:,}', CAST(ROUND((sum(total_compressed_size) / (1024 * 1024))) AS INT)) as compressed_MB,
       format('{:,}', CAST(ROUND((sum(total_uncompressed_size) / (1024 * 1024))) AS INT)) as uncompressed_MB,
       sum(total_compressed_size) / (1024 * 1024 * 1024) as compressed_GB,
       sum(total_uncompressed_size) / (1024 * 1024 * 1024) as uncompressed_GB    
FROM parquet_metadata('data/taxi_2019_04.parquet')

compressed_KB uncompressed_KB compressed_MB uncompressed_MB  compressed_GB  \
0       192,150         474,784           188             464       0.183249   

   uncompressed_GB  
0         0.452789

### Do not show any Pandas dataframe numbers with scientific notation


In [19]:
pd.set_option('display.float_format', '{:.2f}'.format)  # Show 2 decimal places

#### Let us look at some aggregates from our data

In [20]:
%%time
%%dql -t show
SELECT format('{:,}', COUNT(*)) AS RowCount, 
       AVG(passenger_count) AS avg_number_of_passengers,
       AVG(trip_distance) AS avg_trip_distance,
       MAX(trip_distance) AS max_trip_distance,
       AVG(fare_amount) AS avg_fare_amount,
       MAX(fare_amount) AS max_fare_amount,
       AVG(tip_amount) AS avg_tip_amount,
       MAX(tip_amount) AS max_tip_amount FROM 'data/taxi_2019_04.parquet';

┌───────────┬──────────────────────┬────────────────────┬───┬─────────────────┬────────────────────┬────────────────┐
│ RowCount  │ avg_number_of_pass…  │ avg_trip_distance  │ … │ max_fare_amount │   avg_tip_amount   │ max_tip_amount │
│  varchar  │        double        │       double       │   │      float      │       double       │     float      │
├───────────┼──────────────────────┼────────────────────┼───┼─────────────────┼────────────────────┼────────────────┤
│ 7,433,139 │    1.573300727996611 │ 2.9980189376583324 │ … │       395839.94 │ 2.2126877947202788 │          440.8 │
├───────────┴──────────────────────┴────────────────────┴───┴─────────────────┴────────────────────┴────────────────┤
│ 1 rows                                                                                        8 columns (6 shown) │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

CPU times: total: 531 ms
Wall time: 106 ms


#### Notice the extreme fare_amount value ($395,839.94)

#### We need to inspect that record. It could be a corrupt record.

In [21]:
%dql SELECT * FROM 'data/taxi_2019_04.parquet' WHERE fare_amount = (SELECT MAX(fare_amount) FROM 'data/taxi_2019_04.parquet');

vendor_id           pickup_at          dropoff_at  passenger_count  \
0         1 2019-04-29 03:45:58 2019-04-29 03:46:33                1   

   trip_distance rate_code_id store_and_fwd_flag  pickup_location_id  \
0           0.00            1                  N                 145   

   dropoff_location_id payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                  145            2    395839.94   0.50     0.50        0.00   

   tolls_amount  improvement_surcharge  total_amount  congestion_surcharge  
0          0.00                   0.30     395841.25                  0.00

#### <i>As we suspected there is a problem with paying that fare_amount for a trip_distance of 0.00 miles. <br>Also notice the pickup_at, dropoff_at timestamps as well as pickup_location_id and dropoff_location_id.</i>

### If you wish to merge duckdb sql queries and python code, you should use the duckdb.sql('...') query method, 

In [22]:
var1 = duckdb.sql(" SELECT COUNT(*) FROM 'data/taxi_2019_04.parquet' WHERE pickup_at BETWEEN '2019-04-10' AND '2019-04-12'");
print(var1)

┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│       535737 │
└──────────────┘



### or use the `%dql -o <variablename>` method.

In [23]:
%dql -o var2 SELECT COUNT(*) FROM 'data/taxi_2019_04.parquet' WHERE pickup_at BETWEEN '2019-04-10' AND '2019-04-12'
print(var2)

   count_star()
0        535737


## More magic_duckdb

#### List available dql types 

In [24]:
%dql --listtypes

['df', 'df_markdown', 'arrow', 'pl', 'describe', 'show', 'relation']

#### List all tables used in the query

In [25]:
%dql --tables  SELECT COUNT(*) As qCount FROM 'data/taxi_2019_04.parquet'

{'data/taxi_2019_04.parquet'}

#### Get the connection created within DQL and use it directly

In [26]:
con = %dql --getcon
display(con.sql("pragma version").df())

library_version   source_id
0          v1.0.0  1f98600c2c

#### Create a table in an existing database `taxi_2019_04.db`

In [27]:
dql_con = duckdb.connect("data/taxi_2019_04.db")
dql_con.sql("CREATE OR REPLACE TABLE xyz as select * from range(100)")

# -co sets the connection to an existing object
%dql -co dql_con
%dql select count(*) from xyz

count_star()
0           100

#### We can also pass parameters to the query

In [28]:
startdate = '2019-04-22'
enddate = '2019-04-25'
%dql -p startdate -p enddate -o outvar SELECT COUNT(*) As qCount FROM 'data/taxi_2019_04.parquet' WHERE pickup_at BETWEEN ? AND ?;
print(outvar)

   qCount
0  688742


### Create a new persistent duckdb database for the nyc_taxi_data


In [29]:
con = duckdb.connect("data/taxi_2019_04_06.db") 
con.close()

#### Create a connection explicitly, and pass it explicitly to the dql connection. <br>dql by default will use the duckdb default connection

#### Create a table from a our taxi_2019_04 Parquet file

In [30]:
con = duckdb.connect("data/taxi_2019_04_06.db")
con.execute("CREATE OR REPLACE TABLE taxi_trips AS (SELECT * FROM 'data/taxi_2019_*.parquet');")

#### The `-co` sets the connection to an existing database object

In [31]:
%dql -co con
%dql SELECT format('{:,}', COUNT(*)) AS count FROM taxi_trips;

count
0  7,433,139

#### We can also convert to pandas dataFrame directly from your duckdb query<br> by adding the ".df()" function
#### Query the first 10 rows of the newly created taxi_trips data. <br> ####
#### <i>Notice that I use duckdb's to pandas dataframe `df()` at the end of the query.</i>


In [32]:
con.sql("SELECT * FROM taxi_trips LIMIT 3").df()

vendor_id           pickup_at          dropoff_at  passenger_count  \
0         1 2019-04-01 00:04:09 2019-04-01 00:06:35                1   
1         1 2019-04-01 00:22:45 2019-04-01 00:25:43                1   
2         1 2019-04-01 00:39:48 2019-04-01 01:19:39                1   

   trip_distance rate_code_id store_and_fwd_flag  pickup_location_id  \
0           0.50            1                  N                 239   
1           0.70            1                  N                 230   
2          10.90            1                  N                  68   

   dropoff_location_id payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                  239            1         4.00   3.00     0.50        1.00   
1                  100            2         4.50   3.00     0.50        0.00   
2                  127            1        36.00   3.00     0.50        7.95   

   tolls_amount  improvement_surcharge  total_amount  congestion_surcharge  
0          0.00                   0.30          8.80                  2.50  
1          0.00                   0.30          8.30                  2.50  
2          0.00                   0.30         47.75                  2.50

## Analyzing Data with DuckDB


### SQL queries using DuckDB:

To run SQL queries in DuckDB we can directly use ".sql", no need to create a connection to ":memory:"

Every DataFrame inside this notebook will be instantly available for DuckDB to make SQL queries against.


In [33]:
%dql SUMMARIZE SELECT * FROM taxi_trips WHERE vendor_id = 1;

column_name column_type                  min  \
0               vendor_id     VARCHAR                    1   
1               pickup_at   TIMESTAMP  2019-04-01 00:00:02   
2              dropoff_at   TIMESTAMP  2019-04-01 00:03:04   
3         passenger_count     TINYINT                    0   
4           trip_distance       FLOAT                  0.0   
5            rate_code_id     VARCHAR                    1   
6      store_and_fwd_flag     VARCHAR                    N   
7      pickup_location_id     INTEGER                    1   
8     dropoff_location_id     INTEGER                    1   
9            payment_type     VARCHAR                    1   
10            fare_amount       FLOAT                  0.0   
11                  extra       FLOAT                -24.6   
12                mta_tax       FLOAT                  0.0   
13             tip_amount       FLOAT                  0.0   
14           tolls_amount       FLOAT                  0.0   
15  improvement_surcharge       FLOAT                  0.0   
16           total_amount       FLOAT                  0.0   
17   congestion_surcharge       FLOAT                  0.0   

                    max  approx_unique                  avg  \
0                     1              1                 None   
1   2019-04-30 23:59:59        1560287                 None   
2   2019-05-01 00:44:09        1551195                 None   
3                     9             10    1.232603365884064   
4                 830.9            585    2.859716962131162   
5                    99              7                 None   
6                     Y              2                 None   
7                   265            259   164.65948093478713   
8                   265            261   162.15362342929927   
9                     4              4                 None   
10            395839.94           1064   12.972935259103291   
11                 20.0             31    2.664833492625465   
12                27.45              6   0.4970391421784062   
13                315.0           2061    2.141661277488673   
14               812.59            644  0.34971369038846206   
15                  0.3              2  0.29988423851277474   
16            395841.25           7503   18.926066993259646   
17                  2.5              2   2.3142015736366184   

                      std                  q25                  q50  \
0                    None                 None                 None   
1                    None                 None                 None   
2                    None                 None                 None   
3      0.7005261643772758                    1                    1   
4      3.7679879608747138   0.9258681737363619   1.5967366302858295   
5                    None                 None                 None   
6                    None                 None                 None   
7       65.98112677392056                  122                  162   
8       69.90138208319284                  111                  162   
9                    None                 None                 None   
10     238.28318797422514                  6.5    9.390238416050233   
11     0.8157413932679649                  2.5                  2.5   
12     0.0421004930035869                  0.5                  0.5   
13       2.63735446382104                  0.0   1.8821561493981573   
14     1.7909052868971276                  0.0                  0.0   
15  0.0058922519370432375  0.30000001192092896  0.30000001192092896   
16      238.4169124901377   11.270219472323015   14.662266347486998   
17     0.6557249164372682                  2.5                  2.5   

                    q75    count  null_percentage  
0                  None  2767475             0.00  
1                  None  2767475             0.00  
2                  None  2767475             0.00  
3                     1  2767475             0.00  
4    2.9293597400197964  27

#### Let us look at the Datatypes in the table

In [34]:
%dql -co con
%dql DESCRIBE taxi_trips;

column_name column_type null   key default extra
0               vendor_id     VARCHAR  YES  None    None  None
1               pickup_at   TIMESTAMP  YES  None    None  None
2              dropoff_at   TIMESTAMP  YES  None    None  None
3         passenger_count     TINYINT  YES  None    None  None
4           trip_distance       FLOAT  YES  None    None  None
5            rate_code_id     VARCHAR  YES  None    None  None
6      store_and_fwd_flag     VARCHAR  YES  None    None  None
7      pickup_location_id     INTEGER  YES  None    None  None
8     dropoff_location_id     INTEGER  YES  None    None  None
9            payment_type     VARCHAR  YES  None    None  None
10            fare_amount       FLOAT  YES  None    None  None
11                  extra       FLOAT  YES  None    None  None
12                mta_tax       FLOAT  YES  None    None  None
13             tip_amount       FLOAT  YES  None    None  None
14           tolls_amount       FLOAT  YES  None    None  None
15  improvement_surcharge       FLOAT  YES  None    None  None
16           total_amount       FLOAT  YES  None    None  None
17   congestion_surcharge       FLOAT  YES  None    None  None

#### Let us test out query speed in the new duckdb table

In [35]:
%%time
%%dql
SELECT format('{:,}', COUNT(*)) AS RowCount, 
       AVG(passenger_count) AS avg_number_of_passengers,
       AVG(trip_distance) AS avg_trip_distance,
       MAX(trip_distance) AS max_trip_distance,
       AVG(fare_amount) AS avg_fare_amount,
       MAX(fare_amount) AS max_fare_amount,
       AVG(tip_amount) AS avg_tip_amount,
       MAX(tip_amount) AS max_tip_amount FROM taxi_trips;

CPU times: total: 172 ms
Wall time: 50.7 ms


RowCount  avg_number_of_passengers  avg_trip_distance  max_trip_distance  \
0  7,433,139                      1.57               3.00             830.90   

   avg_fare_amount  max_fare_amount  avg_tip_amount  max_tip_amount  
0            13.19        395839.94            2.21          440.80

#### END OF FILE
